File: model/src/main/java/no/nav/pensjon/vtp/testmodell/enheter/EnheterIndeks.java
package no.nav.pensjon.vtp.testmodell.enheter;

import java.util.*;

import static java.util.stream.Collectors.toSet;

public class EnheterIndeks {

    private Map<String, Norg2Modell> byDiskresjonskode = new HashMap<>();

    public void leggTil(Collection<Norg2Modell> enheter) {
        enheter.forEach(m -> byDiskresjonskode.putIfAbsent(m.getDiskresjonskode(), m));
    }

    public Norg2Modell finnByDiskresjonskode(String diskresjonskode) {
       return byDiskresjonskode.get(diskresjonskode);
    }

    public Collection<Norg2Modell> getAlleEnheter() {
        return Collections.unmodifiableCollection(byDiskresjonskode.values());
    }

    public Set<Norg2Modell> findByEnhetIdIn(Collection<String> enhetIds) {
        return enhetIds
                .stream()
                .map(this::finnByEnhetId)
                .flatMap(Optional::stream)
                .collect(toSet());
    }

    public Optional<Norg2Modell> finnByEnhetId(String enhetId) {
        return byDiskresjonskode.values()
                .stream()
                .filter(enhet -> Objects.equals(enhetId, enhet.getEnhetId()))
                .findFirst();
    }
}

File: model/src/main/java/no/nav/pensjon/vtp/testmodell/enheter/EnheterIndeks.kt
package no.nav.pensjon.vtp.testmodell.enheter

import java.util.*
import java.util.function.Consumer

class EnheterIndeks(private val byDiskresjonskode: MutableMap<String, Norg2Modell> = HashMap()) {
    fun saveAll(enheter: Collection<Norg2Modell>) {
        enheter.forEach(Consumer { m: Norg2Modell -> byDiskresjonskode.putIfAbsent(m.diskresjonskode, m) })
    }

    fun finnByDiskresjonskode(diskresjonskode: String): Norg2Modell? {
        return byDiskresjonskode[diskresjonskode]
    }

    val alleEnheter: Collection<Norg2Modell>
        get() = byDiskresjonskode.values

    fun findByEnhetIdIn(enhetIds: Collection<String>): Set<Norg2Modell> {
        return enhetIds
            .mapNotNull { finnByEnhetId(it) }
            .toSet()
    }

    fun finnByEnhetId(enhetId: String): Norg2Modell? {
        return byDiskresjonskode.values.firstOrNull { enhetId == it.enhetId }
    }
}
